### PR TITLE
support concurrent queries using read/write lock and pthreads

### PR DESCRIPTION
**Pull request by swilly22**

### CODE REVIEW

File: src/bulk_insert/bulk_insert_context.c [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+(
+    RedisModuleCtx *ctx,
+    RedisModuleBlockedClient *bc,
+    RedisModuleString **argv,
+    int argc
+) {
+    BulkInsertContext *context = malloc(sizeof(BulkInsertContext));
+
+    context->bc = bc;
+    context->argv = argv;
```
Comment (MeirShpilraien) [inline src/bulk_insert/bulk_insert_context.c]: I think this is not needed right?

File: src/bulk_insert/bulk_insert_context.c [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+void BulkInsertContext_Free
+(
+    BulkInsertContext* ctx
+) {
+    for(int i = 0; i < ctx->argc; i++)
+        RedisModule_Free(ctx->argv[i]);
+    free(ctx->argv);
+    free(ctx);
```
Comment (MeirShpilraien) [inline src/bulk_insert/bulk_insert_context.c]: I am not sure but does the blocked client also need to be freed?

File: src/module.c [hunk 3]
```diff
@@ -62,75 +110,143 @@ Graph *_MGraph_CreateGraph(RedisModuleCtx *ctx, RedisModuleString *graph_name) {
...
+
+    // Force graph pendding operations to complete.
+    Graph_CommitPendingOps(g);
+
+    _MGraph_ReleaseLock();
+
+    // Replay to caller.
+    double t = simple_toc(context->tic);
+    char timings[1024] = {0};
+    sprintf(timings, "%zu Nodes created, %zu Edges created, time: %.6f sec\n", nodes, edges, t);
```
Comment (MeirShpilraien) [inline src/module.c]: its safer to use snprintf

File: src/module.c [hunk 4]
```diff
@@ -43,6 +49,48 @@
...
+
+// TODO: Move lock functions to a suitable location.
+/* Acquire lock for read, multiple threads might be reading concurrently. */
+void _MGraph_AcquireReadLock() {
+    pthread_rwlock_rdlock(&_rwlock);
+}
+
+/* Acquire lock for write, only a single thread can perform work
+ * while holding a write lock. */
+void _MGraph_AcquireWriteLock() {
```
Comment (MeirShpilraien) [inline src/module.c]: Also need to lock the GIL

File: src/module.c [hunk 5]
```diff
@@ -62,75 +110,143 @@ Graph *_MGraph_CreateGraph(RedisModuleCtx *ctx, RedisModuleString *graph_name) {
...
     if (AST_Validate(ast, &reason) != AST_VALID) {
         RedisModule_ReplyWithError(ctx, reason);
         free(reason);
-        return REDISMODULE_OK;
+        return;
     }
 
+    // If this is a write query, acquire write lock.
+    if(AST_ReadOnly(ast)) _MGraph_AcquireWriteLock();
+    else _MGraph_AcquireReadLock();
```
Comment (jeffreylovitz) [inline src/module.c]: I think we have to flip this logic.

File: src/bulk_insert/bulk_insert_context.c (hunk 2 above)
Comment (swilly22) [reply src/bulk_insert/bulk_insert_context.c]: No need

**Pull request merged**